crates/ra_hir/src/ty/infer/unify.rs

```diff
@@ -0,0 +1,122 @@
+//! Unification and canonicalization logic.
+
+use crate::db::HirDatabase;
+use crate::ty::{Ty, Canonical, TraitRef, InferTy};
+use super::InferenceContext;
+
+impl<'a, D: HirDatabase> InferenceContext<'a, D> {
```

Comment:
Unrelated to PR, but I wonder if we should change the definition of inference ctx from

```
struct InferenceContext<'a, D: HirDatabase> {
    db: &'a D,
}
```

to

```
struct InferenceContext<DB> {
    db: DB,
}
```

?

Reply:
Or perhaps it is related to this PR in a sense that we might change `Canonicalizer` and `ChalkContext` to this pattern of hiding lifetime behind generic as well?

Anyway, even if this is a good idea, it's probably better to merge as is and apply refactoiring in a follow-up

Reply:
Hm... does `D: HirDatabase => &D: HirDatabase`?

Reply:
No, but you don't need it, you can place the bound on impl.

See how `DefCollector` is arranged: https://github.com/rust-analyzer/rust-analyzer/blob/9c49f6c36e1e097f938946811d1e2f5eb70edca9/crates/ra_hir/src/nameres/collector.rs#L80-L96

Reply:
Aah :slightly_smiling_face: